relations/src/r1cs/constraint_system.rs

```diff
@@ -149,6 +167,18 @@ impl<F: Field> ConstraintSystem<F> {
         self.mode == SynthesisMode::Setup
     }
 
+    /// Check whether this constraint system aims to optimize weight,
+    /// number of constraints, or neither.
+    pub fn optimization_goal(&self) -> OptimizationGoal {
+        self.optimization_goal
+    }
+
+    /// Specify whether this constraint system should aim to optimize weight,
+    /// number of constraints, or neither.
+    pub fn set_optimization_goal(&mut self, goal: OptimizationGoal) {
+        self.optimization_goal = goal;
```

Comment:
This API makes it seem to me that a circuit can flip-flop to having different optimization goals at different points in circuit generation. Is this intentional?

If not, perhaps the goal should be added at CS initialization, and only have a getter?

Reply:
You are right. Actually this needs to be fixed for the circuit, i.e., can be set only once, before any constraints or variables are generated.

My proposal is to enforce that `set_optimization_goal` would panic if there is any constraint or variable in the constraint system.

Reply:
Putting it as the initialization is a little hard,  for two reasons: (1) we would better not to change the process of creating a cs since it would be overly breaking, yet Rust does not allow a default/optional argument; (2) there is more than one way to create a cs (e.g. by ref).

Reply:
In the long term my idea was to make ConstraintSystem private somehow, and only have ConstraintSystemRef. Then, on ConstraintSystemRef we can have a new constructor “with_optimization_goal” which takes an input a OptimizationGoal and outputs a ConstraintSystemRef with that goal

Reply:
That is a good direction. Then, I think, for this PR, we can keep the current `set_optimization_goal` function for now (which I assume will only be used in Marlin). In the future, Marlin would need to call that new constructor, and setting an optimization goal would only be permitted during the initialization.

Making ConstraintSystem private seems to be a larger and breaking change.

Reply:
This sounds good to me, but can we add a check to ensure that it doesn't change mid circuit generation?

Reply:
Like enforcing (assert) the following?

            num_instance_variables: 1,
            num_witness_variables: 0,
            num_constraints: 0,
            num_linear_combinations: 0,

Reply:
Yeah!